Additional info to debug AccessDenied problems

hubertp opened the pull request:
### Pull Request Description

Plain `Storage failure [AccessDenied].` was rather uninformative when it comes to debugging the underlying problem.
Added more detailed error messages and runners' failures should now sometimes print a detailed message.
References #10662.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.

hubertp committed:
Additional info to debug AccessDenied problems

Plain `Storage failure [AccessDenied].` was rather uninformative when it
comes to debugging the underlying problem.
Added more detailed error messages and runners' failures should now
sometimes print a detailed message.

radeusgd commented on lib/scala/runtime-version-manager/src/main/scala/org/enso/runtimeversionmanager/runner/Command.scala:
```diff
@@ -18,13 +20,25 @@ case class Command(command: Seq[String], extraEnv: Seq[(String, String)]) {
-  def run(): Try[Int] =
+  def run(inheritStdOutErr: Boolean = true): Try[(Int, String)] =
     wrapError {
       logger.debug("Executing {}", this)
       val processBuilder = builder()
-      processBuilder.inheritIO()
-      val process = processBuilder.start()
-      process.waitFor()
+      if (inheritStdOutErr) {
+        processBuilder.inheritIO()
+        val process = processBuilder.start()
+        (process.waitFor(), "")
+      } else {
+        val process = processBuilder.start()
+        val outReader =
+          new BufferedReader(new InputStreamReader(process.getInputStream))
+        val errReader =
+          new BufferedReader(new InputStreamReader(process.getErrorStream))
+        val joiner = new StringJoiner(System.getProperty("line.separator"))
+        outReader.lines.iterator.forEachRemaining(joiner.add)
+        errReader.lines.iterator.forEachRemaining(joiner.add)
+        (process.waitFor(), joiner.toString)
+      }
     }
```
I'm not a fan of this API. In most invocations we discard the returned `""`...

Could we please keep `run(): Try[Int]` as is and just add a separate one `runCapturingOutput(): Try[(Int, String)]` ?

radeusgd replied:
Also keep in mind that this will likely get a race condition if the amount of output is larger - we are reading the streams _after_ `waitFor` exits. But if the internal system buffers get full, the child process trying to write to the streams will block waiting for us to read something - and we will keep waiting for child process to exit - deadlock.

We have to keep reading the buffers in some background thread to avoid this lock scenario.
One implementation of that is in `lib/scala/testkit/src/main/scala/org/enso/testkit/process/WrappedProcess.scala`, you could move it to a non-test library and use it here.

I see even our `System::createProcess` implementation is only partially correct - it reads the streams before `waitFor` but it does so sequentially - but the only correct solution is to read them concurrently, either through Threads or `select`-like multiplexing - as it is possible that e.g. the stderr stream will get saturated while the stdout is not closed yet.

hubertp replied:
Sure I can change the API.

Re WrappedProcess: I can see that you've added quite an elaborate logic to handle streams. Sure, I can use that. Thanks for the pointer.

hubertp committed:
Address PR review

Move WrappedProcess logic out of `testkit` so that it can be used in a
non-test setup.

hubertp committed:
fix imports

JaroslavTulach commented on build.sbt:
```diff
@@ -3189,10 +3191,17 @@ lazy val `runtime-version-manager` = project
   .dependsOn(`distribution-manager`)
 
+lazy val `process-utils` = project
+  .in(file("lib/scala/process-utils"))
```
Are we planning to use this code in production or just in tests? If it is intended for production, I'd rather place the project to `lib/java` and write it in Java.

hubertp replied:
The PR **moves** the file from tests to regular code. It's a plain utility code, please don't ask me to rewrite this to Java at this point.

JaroslavTulach commented on lib/scala/process-utils/src/main/scala/org/enso/process/WrappedProcess.scala:
API should not be created by copying implementation and claiming it is an API now. It deserves some more thoughts and a cleanup. Ideally I'd rather review API in Java (with a Javadoc) rather than Scala code.

JaroslavTulach commented on build.sbt:
```diff
@@ -3189,10 +3191,17 @@ lazy val `runtime-version-manager` = project
   .dependsOn(`edition-updater`)
   .dependsOn(`distribution-manager`)
 
+lazy val `process-utils` = project
```
I am starting to like documentation comments on top of the project describing its purpose.

JaroslavTulach commented on build.sbt:
```diff
@@ -3137,6 +3138,7 @@ lazy val `library-manager-test` = project
     )
   )
   .dependsOn(`library-manager`)
+  .dependsOn(`process-utils`)
```
Shouldn't this be just a `Test` dependency?

hubertp replied:
No, this is correct.

JaroslavTulach reviewed (approved):
Various ramblings, but I guess it is just a refactoring.

radeusgd commented on lib/scala/runtime-version-manager/src/main/scala/org/enso/runtimeversionmanager/runner/Command.scala:
```diff
@@ -27,6 +28,23 @@ case class Command(command: Seq[String], extraEnv: Seq[(String, String)]) {
+      val process        = processBuilder.start()
+      val wrappedProcess = new WrappedProcess(command, process)
+      val stringBuilder  = new StringBuilder()
+      wrappedProcess.appendToBuilder(stringBuilder)
```
The method name `appendToBuilder` is misleading - it sounds like it adds 'what it has currently stored' to the builder and nothing more - which makes this code look incorrect.

But the code is correct, just the name suggests something else.

Could you please rename it? I'd suggest `registerStringBuilderAppendTarget` or `registerStringBuilderForFutureAppends` or something like that.

radeusgd commented on lib/scala/runtime-version-manager/src/main/scala/org/enso/runtimeversionmanager/runner/Command.scala:
```diff
@@ -27,6 +28,23 @@ case class Command(command: Seq[String], extraEnv: Seq[(String, String)]) {
+      val wrappedProcess = new WrappedProcess(command, process)
+      val stringBuilder  = new StringBuilder()
+      wrappedProcess.appendToBuilder(stringBuilder)
+      val result = wrappedProcess.join(waitForDescendants = true, 5 * 60)
```
I'm not sure if that timeout is enough. The timeout was primarily needed in tests. But in the runner, depending on the operation the runtime can be very large:
1. a REPL can stay open for hours
2. a script can keep executing for a long time
3. a dependency preinstall can take a long time over weak connection / large deps.
All these should not have a timeout as their runtime is unbounded.

I guess only `new project` creation is something that is expected to be rather fast and could indeed be OK with a timeout.

I'd suggest to move the timeout parameter to be a parameter of `runAndCaptureOutput` so that we can only set it in places where we know it is OK. Plus, let's make it an `Option` so that one can set no timeout at all.
 As long as it is implicit like currently, it will get forgotten at some point.

hubertp replied:
If something runs for more than 5 minutes then it is bound to be broken. We shouldn't use this for long running operations.

radeusgd replied:
As I told above - there is many 'somethings' that may very well run for more than 5 minutes and that is not an error. Only short operations like new project are indeed small enough that running for this long is definitely an error.

That's why I wanted to make the distinction clear. Currently at call-site it is completely not clear that `run` has no time limit but `runAndCaptureOutput` does have one. That's why I think we should put the timeout in the signature to make that clearer. Or at least add a note in doc comment.

radeusgd replied:
To better explain my worry about this - looking at the API the only difference between `run` and `runAndCaptureOutput` seems to be the 'capturing stdout' aspect. But that is **not true** - they also differ by the fact that one of them offers unbounded execution time whereas the other will timeout. This is not clear from the API. Thus I'm worried that in the future someone may replace a `run` method call with `runAndCaptureOutput` and not realize that this also implies limiting the run-time of the underlying command. As we **do have** long running commands in there, this may be incorrect. And it is a kind of thing that tests will not discover as indeed we don't have any tests that check that the thing will run OK for a longer time.

hubertp committed:
nit
